Cql/Cql.Compiler/ExpressionBuilder.Query.cs

```diff
@@ -86,339 +95,370 @@ protected Expression SingleSourceQuery(Query query)
+                        return lineText;
+                    })
+                    .ToArray();
+            return lines;
+        }
+
+        _logger.LogDebug(
+            """
+            Found {queryType} Query with {sourceCount} source(s) at: {at}
+            Sources:{sources}
+            CQL: {lines}
+            """,
+            ((ReadOnlySpan<string>)["Empty", "Single", "Multi"])[Math.Clamp(sourceLength, 0, 2)],
+            sourceLength,
+            DebuggerView,
+            $"{string.Concat(from s in sources select $"\n\t{s.alias}: {(s.isEnumerationType ? "Enumeration" : "Singleton")} of {s.sourceType}")}",
+            lines is not null ? $"{string.Concat(from l in lines select $"\n\t{l}")}" : "");
+    }
+
+    protected Expression TryDemoteSource(
+        Query query,
+        Expression source,
+        bool isSourcePromoted)
+    {
+        // Because we promoted the source to a list, we now have to demote the result again.
+        if (isSourcePromoted)
+        {
+            var callSingle = _operatorBinding.Bind(CqlOperator.Single, LibraryDefinitionsBuilder.ContextParameter, source);
+            source = callSingle;
+
+        }
+
+        if (query.resultTypeSpecifier is Elm.ListTypeSpecifier && !IsOrImplementsIEnumerableOfT(source.Type))
+        {
+            source = Expression.NewArrayInit(source.Type, source);
         }
 
+        return source;
+    }
+
+    private (Expression sourceExpression, bool[] isSourcePromoted) ProcessQuerySources(Query query)
+    {
+        AliasedQuerySource[] sources = query.source;
+
+        if (sources.Length is 0)
+            throw this.NewExpressionBuildingException("A query must have at least one source.");
+
+        var aliases = sources.SelectToArray(s => s.alias);
+        if (aliases.Any(alias => string.IsNullOrEmpty(alias)))
+            throw this.NewExpressionBuildingException("Query sources must have aliases.");
+
+        var sourceExpressions = sources
+            .SelectToArray(source => TranslateExpression(source.expression));
+
+        // Returns a CrossJoin between IEnumerable<> of T1, T2, T3, etc and return into IEnumerable<(T1, T2, T3, etc)>
+        // a) If a source is not of a list-type (ie, a singleton), it needs to be promoted to a list type.
+        // b) Cross-Join
+        //    IEnumerable<A> a = ...;
+        //    IEnumerable<B> b = ...;
+        //    IEnumerable<c> c = ...;
+        //    IEnumerable<(A, B, C)> crossJoinedValueTupleResults = CrossJoin<A, B, C>(a, b, c);
+
+        var tryPromoteSourceExpressions = sourceExpressions.SelectToArray(expr => TryPromoteSource(expr));
+        var promotedSourceExpressions = tryPromoteSourceExpressions.SelectToArray(s => s.source);
+        var isSourcesPromoted = tryPromoteSourceExpressions.SelectToArray(s => s.isSourcePromoted);
+
+        if (promotedSourceExpressions.Any(promotedSource => !IsOrImplementsIEnumerableOfT(promotedSource.Type)))
```

Comment:
Based on the lines above, how can any of the sources still not be promoted?

Reply:
I changed the wording here to indicate if the original sources were singletons.